### PR TITLE
RaceConfigScenario extension

### PR DESCRIPTION
**Pull request by riidefi**

### CODE REVIEW

**Review by GnomedDev — CHANGES_REQUESTED** [review]

File: payload/game/system/RaceConfig.hh [hunk 1]
```diff
@@ -39,20 +69,8 @@ public:
...
-
-    enum class GameMode {
-        OfflineVS = 1,
-        TimeAttack = 2,
-        OfflineBT = 3,
-        Mission = 4,
-        OnlinePrivateVS = 7,
-        OnlinePrivateBT = 10,
-        Awards = 11,
+    struct Ext {
```
[the remark this answers is not in the document]
Comment (vabold) [reply payload/game/system/RaceConfig.hh:72]: I second `SPScenario`.

File: payload/game/system/RaceConfig.hh [hunk 2]
```diff
@@ -133,6 +161,13 @@ private:
...
     Scenario m_menuScenario;
     Scenario m_awardsScenario;
+
+public:
+    Ext m_spRace;
+    Ext m_spMenu;
+    Ext m_spAwards;
+
+private:
     u8 m_ghostBuffers[2][11][0x2800]; // Modified
```
Comment (vabold) [inline payload/game/system/RaceConfig.hh:171]: We've ensured that it's safe to move the offset of these ghost buffers, right?

**Review by vabold — CHANGES_REQUESTED** [review]

File: payload/game/system/RaceConfig.hh (hunk 2 above)
Comment (riidefi) [reply payload/game/system/RaceConfig.hh:171]: Yes. For readability and cache locality reasons it's also good practice to keep important data clustered at the top.

File: payload/game/system/RaceConfig.hh (hunk 1 above)
[the remark this answers is not in the document]
Comment (riidefi) [reply payload/game/system/RaceConfig.hh:72]: I agree.

File: payload/game/system/RaceConfig.hh [hunk 3]
```diff
@@ -123,6 +141,16 @@ public:
     bool selectRandomCourse();
     void endRace();
 
+    REPLACE void initRaceScenario();
+    REPLACE void initRaceScenarioForAwards();
+    REPLACE void initRaceScenarioForCredits();
```
[the remark this answers is not in the document]
Comment (riidefi) [reply payload/game/system/RaceConfig.hh]: Sure, done.

File: payload/game/system/RaceConfig.hh [hunk 4]
```diff
@@ -133,6 +161,13 @@ private:
     Scenario m_raceScenario;
     Scenario m_menuScenario;
     Scenario m_awardsScenario;
+
+public:
+    Ext m_spRace;
+    Ext m_spMenu;
+    Ext m_spAwards;
```
[the remark this answers is not in the document]
Comment (riidefi) [reply payload/game/system/RaceConfig.hh]: I've made the original Scenarios public too.

File: payload/game/system/RaceConfig.hh [hunk 5]
```diff
@@ -4,12 +4,42 @@
...
+};
+
+constexpr bool IsLegacyOnline(GameMode mode) {
+    switch (mode) {
+    case GameMode::OnlinePrivateVS... GameMode::OnlinePrivateBT:
+        return true;
+    default:
+        return false;
+    }
+}
```
[the remark this answers is not in the document]
Comment (riidefi) [reply payload/game/system/RaceConfig.hh:32]: Do you want me to make another PR for just these? It seemed reasonable to group these together.

[the remark this answers is not in the document]
Comment (GnomedDev) [reply payload/game/system/RaceConfig.hh:32]: I do want you to split it, we shouldn't set the precedent that it's okay to group unrelated changes together.

**Review by GnomedDev — APPROVED** [review]
Once the unrelated changes are split off, LGTM.

File: payload/game/system/RaceConfig.hh (hunk 5 above)
[the remark this answers is not in the document]
Comment (riidefi) [reply payload/game/system/RaceConfig.hh]: Okay, sure, done.

File: payload/game/system/RaceConfig.cc [hunk 6]
```diff
@@ -128,6 +128,19 @@ void RaceConfig::applyCPUMode() {
...
+}
+void RaceConfig::initAwards() {
+    REPLACED(initAwards)();
+    m_spRace = m_spMenu;
+}
+void RaceConfig::initCredits() {
+    REPLACED(initCredits)();
+    m_spRace = m_spMenu;
+}
+
```
Comment (vabold) [inline payload/game/system/RaceConfig.cc:143]:
```suggestion
void RaceConfig::initRace() {
    REPLACED(initRace)();
    m_spRace = m_spMenu;
}

void RaceConfig::initAwards() {
    REPLACED(initAwards)();
    m_spRace = m_spMenu;
    m_spAwards = m_spRace;
}

void RaceConfig::initCredits() {
    REPLACED(initCredits)();
    m_spMenu = m_spAwards;
    m_spRace = m_spMenu;
}
```

This more accurately reflects the base game, so we should likely mirror it with our extended scenarios.

**Review by vabold — CHANGES_REQUESTED** [review]

File: payload/game/system/RaceConfig.cc (hunk 6 above)
Comment (riidefi) [reply payload/game/system/RaceConfig.cc]: Sure, done

**Review by vabold — APPROVED** [review]

**Pull request merged**